File: frontend/src/api/finance.ts
import axios from "axios";
import { useAuth } from "../hooks/useAuth";

const API = axios.create({ baseURL: import.meta.env.VITE_API_URL as string });

API.interceptors.response.use(
  (response) => response,
  (error) => {
    if (error.response?.status === 401) {
      useAuth.getState().logout();
    }
    return Promise.reject(error);
  }
);

const authHeaders = (token: string) => ({
  headers: { Authorization: `Bearer ${token}` },
});

export interface Transaction {
  id: number;
  type: "Income" | "Expense";
  category: string;
  amount: number;
  description: string;
  created_at: string;
}

export interface SummaryItem {
  category: string;
  total: number;
}

export interface AddIncomePayload {
  category_id: number;
  amount: number;
  description?: string;
}

export interface AddExpensePayload {
  category_id: number;
  amount: number;
  description?: string;
}

export type Period = "day" | "week" | "month" | "year" | "all";

const periodParam = (period: Period) =>
  period !== "all" ? `?period=${period}` : "";

export const getBalance = (token: string) =>
  API.get<{ balance: number }>("/finance/balance", authHeaders(token));

export const getTransactions = (token: string, period: Period = "all") =>
  API.get<Transaction[]>(`/finance/transactions${periodParam(period)}`, authHeaders(token));

export const getIncomeSummary = (token: string, period: Period = "all") =>
  API.get<SummaryItem[]>(`/finance/summary/incomes${periodParam(period)}`, authHeaders(token));

export const getExpenseSummary = (token: string, period: Period = "all") =>
  API.get<SummaryItem[]>(`/finance/summary/expenses${periodParam(period)}`, authHeaders(token));

export const addIncome = (token: string, data: AddIncomePayload) =>
  API.post("/finance/incomes", data, authHeaders(token));

export const addExpense = (token: string, data: AddExpensePayload) =>
  API.post("/finance/expenses", data, authHeaders(token));

export const deleteTransaction = (token: string, txType: string, txId: number) =>
  API.delete(`/finance/transactions/${txType}/${txId}`, authHeaders(token));

export interface Category {
  id: number;
  name: string;
}

export const getIncomeCategories = (token: string) =>
  API.get<Category[]>("/finance/categories/incomes", authHeaders(token));

export const getExpenseCategories = (token: string) =>
  API.get<Category[]>("/finance/categories/expenses", authHeaders(token));

File: backend/routers/finance.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from dependencies import get_current_user, get_db
from models.finance import ExpenseCategory, IncomeCategory
from models.user import User
from schemas.finance import (
    BalanceResponse,
    CategoryOut,
    ExpenseCreate,
    IncomeCreate,
    SummaryItem,
    TransactionOut,
)
from services.finance_service import FinanceService

router = APIRouter()


@router.get("/categories/incomes", response_model=List[CategoryOut])
def list_income_categories(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    cats = db.query(IncomeCategory).order_by(IncomeCategory.CategoryID).all()
    return [CategoryOut(id=c.CategoryID, name=c.CategoryName) for c in cats]


@router.get("/categories/expenses", response_model=List[CategoryOut])
def list_expense_categories(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    cats = db.query(ExpenseCategory).order_by(ExpenseCategory.CategoryID).all()
    return [CategoryOut(id=c.CategoryID, name=c.CategoryName) for c in cats]


@router.get("/balance", response_model=BalanceResponse)
def get_balance(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    svc = FinanceService(db)
    return BalanceResponse(balance=svc.calculate_net_balance(current_user.UserID))


@router.get("/transactions", response_model=List[TransactionOut])
def list_transactions(
    limit: int = 20,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    svc = FinanceService(db)
    return svc.get_recent_transactions(current_user.UserID, limit=limit)


@router.post("/incomes", status_code=status.HTTP_201_CREATED)
def add_income(
    data: IncomeCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    svc = FinanceService(db)
    record = svc.add_income(
        user_id=current_user.UserID,
        category_id=data.category_id,
        amount=data.amount,
        description=data.description,
    )
    return {"message": "Income added", "id": record.IncomeID}


@router.post("/expenses", status_code=status.HTTP_201_CREATED)
def add_expense(
    data: ExpenseCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    svc = FinanceService(db)
    current_balance = svc.calculate_net_balance(current_user.UserID)
    if current_balance < data.amount:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Balance cannot go below $0.",
        )
    record = svc.add_expense(
        user_id=current_user.UserID,
        category_id=data.category_id,
        amount=data.amount,
        description=data.description,
    )
    return {"message": "Expense added", "id": record.ExpenseID}


@router.delete("/transactions/{tx_type}/{tx_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_transaction(
    tx_type: str,
    tx_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if tx_type not in ("Income", "Expense"):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid transaction type.")
    svc = FinanceService(db)
    svc.delete_transaction(tx_type, tx_id)


@router.get("/summary/incomes", response_model=List[SummaryItem])
def income_summary(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    svc = FinanceService(db)
    return [
        SummaryItem(category=cat, total=total)
        for total, cat in svc.get_income_summary(current_user.UserID)
    ]


@router.get("/summary/expenses", response_model=List[SummaryItem])
def expense_summary(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    svc = FinanceService(db)
    return [
        SummaryItem(category=cat, total=total)
        for total, cat in svc.get_expense_summary(current_user.UserID)
    ]
